File: Juan/P4/Ruby/Napakalaki/lib/Monster.rb
#encoding: utf-8

#Versión 3.0

require_relative 'TreasureKind.rb'
require_relative 'BadConsequence.rb'
require_relative 'Prize.rb'



module NapakalakiGame
class Monster
    # @@numMonster --> contar el número de monstruos
    #Atributos propios de la clase
  
    # @name # String para representar el nombre del monstruo
    # @combatLevel # int para representar el nivel de combate del monstruo
  
    #Otros atributos
    # @badconsequence  # de tipo BadConsequence
    # @price # de tipo Prize
 
  
  
    def initialize(name,level,bc,price,ic)
        @name = name
        @combatLevel = level    
        @prize = price
        @badconsequence = bc
        @levelChangeAgainstCultistPlayer = ic
    
    end
    
#    private_class_method :new
#    def self.oldMonster(name, level,bc,price)
#        new(name,level,bc,price,0)
#    end
#    
#    def self.newMonster(name, level,bc,price,ic)
#        new(name,level,bc,price,ic)
#    end
  
    ## Métodos get
    attr_reader :name
    attr_reader :combatLevel 
    attr_reader :badconsequence
    attr_reader :prize
    
   
    
    def getLevelsGained
        @prize.levels
    end
    
    def getTreasureGained
        @prize.treasures
    end
    
    def getCombatLevelAgainstCultistPlayer
        return @levelChangeAgainstCultistPlayer + @combatLevel
    end

    def to_s
       return "Monster =  #{@name}  , combatLevel =   #{@combatLevel} , bc = #{@badconsequence} ,  Price:  #{@prize} , levelChanged: #{@ic}"      
    end
   
   
end
end

File: Juan/P5/java/Napakalaki/src/NapakalakiGame/Monster.java
package NapakalakiGame;

import java.net.URL;

/**
 *
 * @author juane
 * @version 2.1
 * Clase Monster
 */
public class Monster {
    //private static numMonster; contar el numero de objetos
    /* Atributos propios de la clase*/
    
    private String name; // nombre del monstruo
    private int combatLevel; // nivel de combate del monstruo
    private int levelChangeAgainstCultistPlayer = 0;

    /* Relaciones atributo de tipo Prize y BadConsequence */
    private Prize prize; 
    private BadConsequence badConsequence;
     
    /* Constructor*/
    
    /**
     * Constructor  Monster(String n, int l, BadConsequence b, Prize p)
     * @param n Nombre del monstruo
     * @param l nivel del monstruo
     * @param b mal royo
     * @param p buen royo
     */
    Monster(String n, int l, BadConsequence b, Prize p){
        
        this.name = n;
        this.combatLevel = l;
        this.badConsequence = b;
        this.prize = p;
        this.levelChangeAgainstCultistPlayer=0;
    }
    /**
     * Constructor  Monster(String n, int l, BadConsequence badconsequence, Prize p, int IC)
     * @param n String  
     * @param l int
     * @param badconsequence Badconsequence
     * @param p Prize
     * @param IC int , incremento y decremento cuando combate con un
     * cultisplayer
     */
    Monster(String n, int l, BadConsequence badconsequence, Prize p, int IC){
        this.name = n;
        this.combatLevel = l;
        this.badConsequence = badconsequence;
        this.prize = p;
        this.levelChangeAgainstCultistPlayer = IC;
    }
    
    /**
     * Método getName()
     * Devuelve el nombre del monstruo
     * @return name de tipo int
     */
    public String getName() {
        return name;
    }

    /**
     * Método getCombatLevel()
     * Devuelve el nivel de combate del monstruo
     * @return 
     */
    public int getCombatLevel() {
        return combatLevel;
    }

    public int getCombatLevelAgainstCultistPlayer() {
        return levelChangeAgainstCultistPlayer + combatLevel;
    }
    

    /**
     * Método getBadConsequence()
     * Devuelve el mal royo
     * @return badConsequence de tipo BadConsequence
     */
    public BadConsequence getBadConsequence(){
        return this.badConsequence; 
    }
    
    /**
     * Método getLevelsGained()
     * Devuelve el número de niveles ganados proporcionados por su buen rollo
     * @return 
     */
    public int getLevelsGained(){         
        
        return this.prize.getLevels(); 
    }
    
    /**
     * Método getTreasuresGained()
     * Devuelve el número de tesoros ganados proporcionados por su buen rollo
     * @return 
     */
    public int getTreasuresGained(){        
        
        return this.prize.getTreasures(); 
    }
    
    
    
    /**
     * Método getPrice()
     * Devuelve el buen royo
     * @return  price de tipo Prize
     */
    /*No está en el diagrama pero deveria estar!!!*/
    public Prize getPrice() {
        return prize;
    }

    public String getIcon(){
        String path ="Source Packages/GUI/Monster/";
   
        
        path = path.concat(getName()+".jpg");
        return path;
    }
    
    
    /**
     * Método toString
     * @return 
     */
    @Override
    public String toString() {
        return "Monster = " + this.name + ", combatLevel = " 
                +  this.combatLevel + ", Price: " +  this.prize 
                + ", " + this.badConsequence ;
    }
}
